arangod/RocksDBEngine/Methods/RocksDBSstFileMethods.cpp

```diff
@@ -0,0 +1,180 @@
+
+#include "RocksDBSstFileMethods.h"
+
+#include "ApplicationFeatures/ApplicationServer.h"
+#include "Basics/files.h"
+#include "Basics/FileUtils.h"
+#include "Basics/RocksDBUtils.h"
+#include "Random/RandomGenerator.h"
+#include "RestServer/DatabasePathFeature.h"
+
+using namespace arangodb;
+namespace FU = basics::FileUtils;
+
+RocksDBSstFileMethods::RocksDBSstFileMethods(
+    bool isForeground, rocksdb::DB* rootDB,
+    RocksDBTransactionCollection* trxColl, RocksDBIndex& ridx,
+    rocksdb::Options const& dbOptions, std::string const& idxPath)
+    : RocksDBMethods(),
+      _isForeground(isForeground),
+      _rootDB(rootDB),
+      _trxColl(trxColl),
+      _ridx(ridx),
+      _sstFileWriter(rocksdb::EnvOptions(dbOptions), dbOptions,
+                     ridx.columnFamily()->GetComparator(), ridx.columnFamily()),
+      _cf(ridx.columnFamily()),
+      _idxPath(idxPath) {}
+
+RocksDBSstFileMethods::~RocksDBSstFileMethods() { cleanUpFiles(); }
+
+void RocksDBSstFileMethods::insertEstimators() {
+  auto ops = _trxColl->stealTrackedIndexOperations();
+  if (!ops.empty()) {
+    TRI_ASSERT(_ridx.hasSelectivityEstimate() && ops.size() == 1);
+    auto it = ops.begin();
+    TRI_ASSERT(_ridx.id() == it->first);
+
+    auto* estimator = _ridx.estimator();
+    if (estimator != nullptr) {
+      if (_isForeground) {
+        estimator->insert(it->second.inserts);
+      } else {
+        uint64_t seq = _rootDB->GetLatestSequenceNumber();
+        // since cuckoo estimator uses a map with seq as key we need to
+        estimator->bufferUpdates(seq, std::move(it->second.inserts),
+                                 std::move(it->second.removals));
+      }
+    }
+  }
+}
+
+rocksdb::Status RocksDBSstFileMethods::writeToFile() {
+  if (_sortedKeyValPairs.empty()) {
+    return rocksdb::Status::OK();
+  }
+
+  auto comparator = _cf->GetComparator();
+  std::sort(_sortedKeyValPairs.begin(), _sortedKeyValPairs.end(),
+            [&comparator](auto& v1, auto& v2) {
+              return comparator->Compare({v1.first}, {v2.first}) < 0;
+            });
+  TRI_pid_t pid = Thread::currentProcessId();
+  std::string tmpFileName =
+      std::to_string(pid) + '-' +
+      std::to_string(RandomGenerator::interval(UINT32_MAX));
+  std::string fileName = FU::buildFilename(_idxPath, tmpFileName);
+  fileName += ".sst";
+  rocksdb::Status res = _sstFileWriter.Open(fileName);
+  if (res.ok()) {
+    _bytesToWriteCount = 0;
+    _sstFileNames.emplace_back(fileName);
+    for (auto const& [key, val] : _sortedKeyValPairs) {
+      res = _sstFileWriter.Put(rocksdb::Slice(key), rocksdb::Slice(val));
+      if (!res.ok()) {
+        break;
+      }
+    }
+    _sortedKeyValPairs.clear();
+    if (res.ok()) {
+      res = _sstFileWriter.Finish();
+    }
+    if (!res.ok()) {
+      cleanUpFiles();
+    } else {
+      insertEstimators();
+    }
+  }
+  return res;
+}
+
+rocksdb::Status RocksDBSstFileMethods::stealFileNames(
```

Comment:
Can we make this method return an `arangodb::Result` instead of a `rocksdb::Status`?
There is not much need for it to return a `rocksdb::Status`, because it is not an overridden method.

Reply:
done